---
title: "Marker Gene Trajectory Inference from Single-Cell RNA-seq Data Using a Semi-Supervised Bayesian Approach"
author: "Junchao Wang, Ling Sun, Yumiao Hou, Naiqian Zhang*"
date: "December 2024"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(BayesTraj)
library(pheatmap)
library(ggplot2)
library(RColorBrewer)

theme_set(theme_bw())
set.seed(1L)
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, warning = FALSE,
                      fig.center = TRUE, fig.width = 6, fig.height = 4)

```

# Introduction

`BayesTraj` is a probabilistic pseudotime framework. BayesTraj

- **A.** Inputs to BayesTraj
- **B.** Illustration of the Bayesian Model
- **C.** Trend of Marker Gene Changes During Differentiation
- **D.** Generation of Pseudo-time and Branch Probabilities
- **E.** Branch-specific gene detection

  ![](https://github.com/user-attachments/assets/557a698d-6c90-410a-aaf8-1045d956a93c)
  


Under the hood, BayesTraj employs Markov Chain Monte Carlo (MCMC) sampling methods to conduct the Bayesian inference using the [Stan](http://mc-stan.org) probabilistic programming language.

In this vignette, we demonstrate three common applications of BayesTraj:

1. **Linear trajectories**: Continuous differentiation, such as endoderm formation from stem cells.
2. **Bifurcation trajectories**: Branching into distinct fates, like AT1/AT2 pneumocytes.
3. **multifurcation trajectories**: Complex processes with multiple branches, such as hematopoiesis.

BayesTraj is benchmarked against state-of-the-art methods using both simulated and real scRNA-seq datasets. It excels in pseudotime accuracy, branch detection, and gene activation analysis, often achieving higher precision and recall. Notably, BayesTraj does not require predefined starting points and handles challenging scenarios, such as imbalanced data or high dropout rates. 


# Application Scenarios


## Linear trajectories


To assess the performance of BayesTraj in inferring real cellular trajectories, we first applied it to scRNA-seq datasets with linear trajectories. 

### Data input

BayesTraj accepts input in the form of a cell-by-gene expression matrix with non-negative values. For optimal performance, we recommend using data normalized as `log2(TPM + 1)` or `log2(RPKM + 1)`, as these formats align with the model's designed mean-variance relationship.

Here, we applied pseudotime inference on a time-series scRNA-seq dataset of mESC, leveraging a small panel of well-characterized marker genes, including early endodermal cell markers (Gata4, Sox17, Foxa2, and Gata6) and pivotal regulatory genes (EOMES, CER1, and PRDM1).

```{r load-hESC-data}
hesc_exp <- read.csv("data/linear/hesc_exp.csv", header = T, row.names = 1)
hesc_exp[1:3,1:5]
gene_names <- rownames(hesc_exp)
t_hesc_exp <- t(hesc_exp)           # Rows represent cells, and columns represent genes.
```

Select known marker genes and extract their corresponding expression matrix.

```{r hESC-marker-genes}
marker_genes1 <- c('EOMES','CER1', 'GATA4', 'DKK4','MYCT1','PRDM1')
marker_genes2 <- c('CDX1','MSX2','T')
marker_exp = hesc_exp[c(marker_genes1,marker_genes2),]
```


Visualize the expression of marker genes using a heatmap to determine whether they exhibit `switch-like` or `transient` expression patterns.

```{r Visualize_hESC_marker_gene}
bluecolors <- brewer.pal(9,"Blues")
redcolors <- brewer.pal(9,"Reds")
color_palette <- c(colorRampPalette(colors = c(bluecolors[9], bluecolors[7]))(20),
                   colorRampPalette(colors = c(bluecolors[7], bluecolors[5]))(20),
                   colorRampPalette(colors = c(bluecolors[5], "white"))(10),
                   colorRampPalette(colors = c("white", redcolors[5]))(10),
                   colorRampPalette(colors = c(redcolors[5], redcolors[6]))(20),
                   colorRampPalette(colors = c(redcolors[6], redcolors[7]))(20))
pheatmap(marker_exp,
         scale = "row",
         color = color_palette,
         cluster_rows = TRUE,
         cluster_cols = F,
         show_rownames = T,
         show_colnames = F
)

```


Normalize the expression data of marker genes.

```{r normalize-expression_hESC}
norm_marker_exp <- normalized_expression_data(t(marker_exp))
Y1 <- norm_marker_exp[,c(marker_genes1),drop = F]
Y2 <- norm_marker_exp[,c(marker_genes2),drop = F]
```

### Fitting with BayesTraj


```{r Fit_hESC}
b1_mu0_s_mu = rep(0.5,length(marker_genes1))
b1_mu0_s_sd = rep(0.2,length(marker_genes1))
b2_mu0_s_mu = rep(0.5,length(c(marker_genes2)))
b2_mu0_s_sd = rep(0.2,length(c(marker_genes2)))

num_genes <- length(unique(c(marker_genes1,marker_genes2)))

b1_mu0_t_mu <- rep(0.25,num_genes-length(marker_genes1))
b1_mu0_t_sd <- rep(0.2,num_genes-length(marker_genes1))

b2_mu0_t_mu <- rep(0.25,num_genes-length(c(marker_genes2)))
b2_mu0_t_sd <- rep(0.2,num_genes-length(c(marker_genes2)))

b1_k_s_mu <- rep(0,length(marker_genes1))
b1_k_s_sd <- rep(50,length(marker_genes2))

bt <- BayesTraj(iter = 100,n_branch = 1,
                            Y1 = Y1,
                            Y2 = Y2,
                            alpha = 0.1,eplison = 0.04,
                            lower_k_s = 5,lower_k_t = 1,lower_t0 = 0.1,upper_b1_t0 = 0.8,
                            lower_b2_t0 = 0.1,
                            b1_mu0_s_mu = b1_mu0_s_mu,b1_mu0_s_sd = b1_mu0_s_sd,
                            # b1_k_s_mu = b1_k_s_mu,b1_k_s_sd = b1_k_s_sd,
                            b2_mu0_s_mu = b2_mu0_s_mu,b2_mu0_s_sd = b2_mu0_s_sd ,
                            b1_mu0_t_mu = b1_mu0_t_mu,b1_mu0_t_sd = b1_mu0_t_sd,
                            b2_mu0_t_mu = b2_mu0_t_mu,b2_mu0_t_sd = b2_mu0_t_sd ,
                            lower_mu0_s = 0.2,upper_mu0_s = 0.5,lower_mu0_t = 0.05,upper_mu0_t=0.4,
                            chains = 1)
```
### Examining results

```{r results}
bt <- infer_pseudotime(bt)
bt <- infer_mu0(bt)
bt <- infer_k(bt)
bt <- infer_t0(bt)
bt <- branch_probability(bt)
bt <- calculate_branches(bt)

```
### Extracting useful quantities

```{r}
label <- rep('esc',nrow(Y1))
color <- c('lightblue')
names(color) <- 'esc'
plot_branches(bt,branch = 1,ncol = 4,label = label,label_color = color)

```

### Branch-specific gene detection

```{r branch_test, message=FALSE}

t <- bt$all_branch_data[[1]]$t
expr <- hesc_exp[,bt$all_branch_data[[1]]$index]
```

```{r results='hide'}
filtered_genes <- filter_de_genes(t,as.matrix(expr),limit_cor = 0.6,limit_bf = 1e40,limit_min_t0 = 0.1,limit_max_t0 = 1,
                limit_min_k = 10,limit_min_mu0 = 2.5)

```

```{r}
genes <- filtered_genes[['selected_genes']]


para <- filtered_genes$bf
ordered_para <- para[order(para[,1],decreasing = T),]
head(ordered_para)

par(mfrow = c(2,2))
for (gene in rownames(ordered_para)[1:12]) {
  plot(t, expr[gene,], main = gene, xlab = "pseudotime", ylab = "expression", pch = 16,col = 'grey20')
  curve(ordered_para[gene,2] / (1 + exp(-ordered_para[gene,3] * (x - ordered_para[gene,4]))), 
        add = TRUE, col = "red", lwd = 2)
}

```


```{r}

para <- filtered_genes$bf
ordered_para <- para[order(para[,1],decreasing = T),]
head(ordered_para)

par(mfrow = c(2,2))
for (gene in rownames(ordered_para)[1:20]) {
  plot(t, expr[gene,], main = gene, xlab = "pseudotime", ylab = "expression", pch = 16,col = 'grey20')
  curve(ordered_para[gene,2] / (1 + exp(-ordered_para[gene,3] * (x - ordered_para[gene,4]))), 
        add = TRUE, col = "red", lwd = 2)
}

# debug(process_and_smooth_data)
expr1 <- process_and_smooth_data(expr,genes,smooth_k = 8)

input_list <- list(expr1)
# debug(generate_heatmap)
generate_heatmap(input_list)


```


## Bifurcation trajectories

We use scRNA-seq datasets with a bifurcation structure to again examine whether BayesTraj can successfully detect a branching trajectory.

### Data Input and Preprocessing

```{r data-input-lung, message=FALSE, warning=FALSE}
library(BayesTraj)

# Load data
data <- read.csv('data/bifurcation/lung_exp.csv', header = T, row.names = 1)
gene_names <- rownames(data)

# Transform data
data_counts <- t(log1p(data))

# Define gene sets for analysis
AT1_genes <- c('Clic5', 'Pdpn', 'Cav1', 'Ager', 'Hopx', 'Vegfa')
AT2_genes <- c('Sftpa1', 'Lyz2', 'Sftpb', 'Sftpc', 'Slc34a2')

# Check if genes exist in the dataset
AT1_genes %in% gene_names
AT2_genes %in% gene_names

# Subset gene expression matrix
gene_expression_matrix <- t(data_counts[, c(AT1_genes, AT2_genes)])
```

---

###  Visualizing Gene Expression with Heatmap

```{r heatmap-lung, message=FALSE, warning=FALSE}


# Load cell labels
cell_labels <- read.csv('data/bifurcation/cell_labels.csv', header = F)$V1
desired_order <- c('cell_type_1', 'cell_type_2', 'cell_type_3')
annotation <- data.frame(Label = factor(cell_labels, levels = desired_order))
rownames(annotation) <- colnames(t(data_counts))

# Sort cells by label
sorted_order <- order(annotation$Label)
gene_expression_matrix <- gene_expression_matrix[, sorted_order]
annotation <- annotation[sorted_order, , drop = FALSE]

# Define color palette
bluecolors <- brewer.pal(9, "Blues")
redcolors <- brewer.pal(9, "Reds")
color_palette <- c(
  colorRampPalette(colors = c(bluecolors[9], bluecolors[7]))(20),
  colorRampPalette(colors = c(bluecolors[7], bluecolors[5]))(20),
  colorRampPalette(colors = c(bluecolors[5], "white"))(10),
  colorRampPalette(colors = c("white", redcolors[5]))(10),
  colorRampPalette(colors = c(redcolors[5], redcolors[6]))(20),
  colorRampPalette(colors = c(redcolors[6], redcolors[7]))(20)
)

# Annotation colors
unique_labels <- levels(annotation$Label)
colors <- c('blue', '#FFFF00', 'green')
names(colors) <- unique_labels
ann_colors <- list(Label = colors)

# Plot heatmap
pheatmap(
  gene_expression_matrix,
  annotation_col = annotation,
  annotation_colors = ann_colors,
  color = color_palette,
  scale = "row",
  cluster_rows = TRUE,
  cluster_cols = F,
  show_rownames = T,
  show_colnames = F
)
```

---

###  Bayesian Modeling with BayesTraj

```{r bayesian-modeling-lung, message=FALSE, warning=FALSE}
# Preprocess data for BayesTraj
sub_data_counts <- normalized_expression_data(data_counts)
sub_data_counts <- sub_data_counts[, c(AT1_genes, AT2_genes)]
Y1 <- sub_data_counts[, c(AT1_genes), drop = F]
Y2 <- sub_data_counts[, c(AT2_genes), drop = F]

AT1_df <- sub_data_counts[cell_labels %in% c('cell_type_2'),]
AT2_df <- sub_data_counts[cell_labels %in% c('cell_type_3'),]
par(mfrow = c(1,1))
boxplot(AT1_df)
boxplot(AT2_df)

quantile_AT1 <- apply(AT1_df,2,function(x){
  quantile(x,probs = c(0.25,0.5,0.75))
})
quantile_AT2 <- apply(AT2_df,2,function(x){
  quantile(x,probs = c(0.25,0.5,0.75))
})
quantile_all <- rbind(quantile_AT1[2,],quantile_AT2[2,])
quantile_all <- quantile_all[,unique(c(AT1_genes,AT2_genes))]
transient_mu <- apply(quantile_all, 2, function(x){
  sort(x, decreasing = TRUE)[2]
})
quantile_all <- rbind(quantile_AT1[3,],quantile_AT2[3,])
quantile_all <- quantile_all[,unique(c(AT1_genes,AT2_genes))]
switch_mu <- apply(quantile_all, 2, function(x){
  sort(x, decreasing = TRUE)[1]
})

b1_mu0_s_mu = switch_mu[AT1_genes]/2
b1_mu0_s_sd = rep(0.05,length(AT1_genes))
b2_mu0_s_mu = switch_mu[c(AT2_genes)]/2
b2_mu0_s_sd = rep(0.05,length(c(AT2_genes)))
num_genes <- length(unique(c(AT1_genes,AT2_genes)))
b1_mu0_t_mu <- transient_mu[!(names(transient_mu) %in% AT1_genes)]/4
b1_mu0_t_sd <- rep(0.02,num_genes-length(AT1_genes))
b2_mu0_t_mu <- transient_mu[!(names(transient_mu) %in% c(AT2_genes))]/4
b2_mu0_t_sd <- rep(0.02,num_genes-length(c(AT2_genes)))

bt <- BayesTraj(iter = 100,n_branch = 2,
                Y1 = Y1,
                Y2 = Y2,
                alpha = 0.1,eplison = 0.04,
                lower_k_s = 10,lower_k_t = 4,lower_t0 = 0.3,upper_b1_t0 = 0.8,lower_b2_t0 = 0.3,
                lower_mu0_s = 0.2,upper_mu0_s = 0.5,lower_mu0_t = 0.05,upper_mu0_t=0.4,

                b1_mu0_s_mu = b1_mu0_s_mu,b1_mu0_s_sd = b1_mu0_s_sd,
                b2_mu0_s_mu = b2_mu0_s_mu,b2_mu0_s_sd = b2_mu0_s_sd ,
                b1_mu0_t_mu = b1_mu0_t_mu,b1_mu0_t_sd = b1_mu0_t_sd,
                b2_mu0_t_mu = b2_mu0_t_mu,b2_mu0_t_sd = b2_mu0_t_sd ,

                chains = 1)

```

---

### Results and Visualization

#### Pseudotime and Branch Inference

```{r pseudotime-inference-lung, message=FALSE, warning=FALSE}
bt <- infer_pseudotime(bt)
bt <- infer_mu0(bt)
bt <- infer_k(bt)
bt <- infer_t0(bt)
bt <- branch_probability(bt)
bt <- calculate_branches(bt, limit_prob = 0.9)
```

#### Branch-Specific Gene Analysis

```{r branch-analysis-lung, message=FALSE, warning=FALSE}
# Analyze branch-specific dynamics
t1 <- bt$all_branch_data[[1]]$t
expr1 <- data[, bt$all_branch_data[[1]]$index]
filtered_genes1 <- filter_de_genes(t1, as.matrix(expr1), limit_cor = 0.4, limit_bf = 1, limit_min_t0 = 0.1)

# Plot top genes for branch 1
para1 <- filtered_genes1$bf
ordered_para1 <- para1[order(para1[, 1], decreasing = T), ]
par(mfrow = c(2, 5))
for (gene in rownames(ordered_para1)[1:10]) {
  plot(t1, expr1[gene, ], main = gene, xlab = "Pseudotime", ylab = "Expression", pch = 16, col = 'grey20')
  curve(ordered_para1[gene, 2] / (1 + exp(-ordered_para1[gene, 3] * (x - ordered_para1[gene, 4]))), add = TRUE, col = "red", lwd = 2)
}
```

---

### Conclusion

This analysis demonstrated the power of BayesTraj for identifying and visualizing branch-specific dynamics in scRNA-seq data. Key genes driving bifurcation events were identified, providing insights into cellular trajectories in lung development.


## multifurcation trajectories

We used a scRNA-seq dataset from the mouse hematopoietic system to further examine whether BayesTraj can reconstruct multiple lineages. Using lineage-specific marker genes for erythroid (Gata1, Klf1, Epor, Itga2b, Pf4, Tal1), granulocytic (S100a8, S100a9, Gfi1, Camp, Cebpe), and monocytic differentiation (Ly86, Irf8, Ccr2, Cx3cr1), we applied BayesTraj to reconstruct a trajectory with two major branches and three distinct fates.

### Data Input and Preprocessing

```{r data-input, message=FALSE, warning=FALSE}
library(BayesTraj)

# Load data
data <- read.csv('D:/BayesTraj/data/multifurcation/data_Olsson_wishbone.csv', row.names = 1)

data_counts <- t(as.matrix(data)) 
gene_names <- rownames(data_counts)
# Define lineage-specific gene sets
ery_genes <- c('Gata1', 'Klf1', 'Epor', 'Itga2b', 'Pf4', 'Tal1')
gran_genes <- c('S100a8', 'S100a9', 'Gfi1', 'Camp', 'Cebpe')
mono_genes <- c('Ly86', 'Irf8', 'Ccr2', 'Cx3cr1')

# Validate genes
ery_genes %in% gene_names
gran_genes %in% gene_names
mono_genes %in% gene_names

# Subset gene expression matrix
gene_expression_matrix <- data_counts[c(ery_genes, gran_genes, mono_genes),]

```

---

### Visualizing Gene Expression with Heatmap

```{r heatmap, message=FALSE, warning=FALSE}


# Load cell labels and sort
cell_labels <- read.csv('D:/BayesTraj/data/multifurcation/cell_label.tsv.gz', sep = '\t', header = F)$V1
desired_order <- c('HSCP-1', 'HSCP-2', 'Multi-Lin', 'Meg', 'Eryth', 'Gran', 'Myelocyte', 'MDP', 'Mono')
annotation <- data.frame(Label = factor(cell_labels, levels = desired_order))
rownames(annotation) <- colnames(data_counts)

sorted_order <- order(annotation$Label)
gene_expression_matrix <- gene_expression_matrix[, sorted_order]
annotation <- annotation[sorted_order, , drop = FALSE]

# Define colors for heatmap
unique_labels <- levels(annotation$Label)
colors <- c('#FEE551', '#E4BE00', '#E84037', '#DB1135', '#9B278A', '#6081C0', '#55B1E4', '#00973D', '#71BD58')
names(colors) <- unique_labels
ann_colors <- list(Label = colors)

bluecolors <- brewer.pal(9, "Blues")
redcolors <- brewer.pal(9, "Reds")
color_palette <- c(
  colorRampPalette(colors = c(bluecolors[9], bluecolors[7]))(20),
  colorRampPalette(colors = c(bluecolors[7], bluecolors[5]))(20),
  colorRampPalette(colors = c(bluecolors[5], "white"))(10),
  colorRampPalette(colors = c("white", redcolors[5]))(10),
  colorRampPalette(colors = c(redcolors[5], redcolors[6]))(20),
  colorRampPalette(colors = c(redcolors[6], redcolors[7]))(20)
)

# Plot heatmap
pheatmap(
  gene_expression_matrix,
  annotation_col = annotation,
  annotation_colors = ann_colors,
  color = color_palette,
  scale = "row",
  cluster_rows = TRUE,
  cluster_cols = F,
  show_rownames = T,
  show_colnames = F
)
```

---

### Bayesian Modeling with BayesTraj

```{r bayesian-modeling, message=FALSE, warning=FALSE}
# Normalize and subset data
sub_data_counts <- normalized_expression_data(data)
sub_data_counts <- sub_data_counts[,c(ery_genes,gran_genes,mono_genes)]
Y1 <- sub_data_counts[,c(ery_genes),drop = F]
Y2 <- sub_data_counts[,c(gran_genes),drop = F]
Y3 <- sub_data_counts[,c(mono_genes),drop = F]

ery_df <- sub_data_counts[cell_labels %in% c('Eryth','Meg'),]
mono_df <- sub_data_counts[cell_labels %in% c('Mono','MDP'),]
gran_df <- sub_data_counts[cell_labels %in% c('Gran','Myelocyte'),]
par(mfrow = c(3,1))
boxplot(ery_df)
boxplot(gran_df)
boxplot(mono_df)


quantile_ery <- apply(ery_df,2,function(x){
  quantile(x,probs = c(0.25,0.5,0.75))
})
quantile_gran <- apply(gran_df,2,function(x){
  quantile(x,probs = c(0.25,0.5,0.75))
})
quantile_mono <- apply(mono_df,2,function(x){
  quantile(x,probs = c(0.25,0.5,0.75))
})

quantile_all <- rbind(quantile_ery[2,],quantile_gran[2,],quantile_mono[2,])
quantile_all <- quantile_all[,unique(c(ery_genes,gran_genes,mono_genes))]

transient_mu <- apply(quantile_all, 2, function(x){
  sort(x, decreasing = TRUE)[2]
})

quantile_all <- rbind(quantile_ery[3,],quantile_gran[3,],quantile_mono[3,])
quantile_all <- quantile_all[,unique(c(ery_genes,gran_genes,mono_genes))]

switch_mu <- apply(quantile_all, 2, function(x){
  sort(x, decreasing = TRUE)[1]
})


b1_mu0_s_mu = switch_mu[ery_genes]/2
b1_mu0_s_sd = rep(0.05,length(ery_genes))
b2_mu0_s_mu = switch_mu[c(gran_genes)]/2
b2_mu0_s_sd = rep(0.05,length(c(gran_genes)))
b3_mu0_s_mu = switch_mu[c(mono_genes)]/2
b3_mu0_s_sd = rep(0.05,length(c(mono_genes)))

num_genes <- length(unique(c(ery_genes,gran_genes,mono_genes)))

b1_mu0_t_mu <- transient_mu[!(names(transient_mu) %in% ery_genes)]/2
b1_mu0_t_sd <- rep(0.02,num_genes-length(ery_genes))

b2_mu0_t_mu <- transient_mu[!(names(transient_mu) %in% c(gran_genes))]/2
b2_mu0_t_sd <- rep(0.02,num_genes-length(c(gran_genes)))

b3_mu0_t_mu <- transient_mu[!(names(transient_mu) %in% c(mono_genes))]/2
b3_mu0_t_sd <- rep(0.02,num_genes-length(c(mono_genes)))


bt <- BayesTraj(iter = 100,n_branch = 3,
                Y1 = Y1,
                Y2 = Y2,
                Y3 = Y3,
                alpha = 0.1,eplison = 0.04,
                lower_k_s = 10,lower_k_t = 4,lower_t0 = 0.3,upper_b1_t0 = 0.5,lower_b2_t0 = 0.5,
                lower_mu0_s = 0.2,upper_mu0_s = 0.5,lower_mu0_t = 0,upper_mu0_t=0.25,
                
                b1_mu0_s_mu = b1_mu0_s_mu,b1_mu0_s_sd = b1_mu0_s_sd,
                b2_mu0_s_mu = b2_mu0_s_mu,b2_mu0_s_sd = b2_mu0_s_sd ,
                b3_mu0_s_mu = b3_mu0_s_mu,b3_mu0_s_sd = b3_mu0_s_sd,
                b1_mu0_t_mu = b1_mu0_t_mu,b1_mu0_t_sd = b1_mu0_t_sd,
                b2_mu0_t_mu = b2_mu0_t_mu,b2_mu0_t_sd = b2_mu0_t_sd ,
                b3_mu0_t_mu = b3_mu0_t_mu,b3_mu0_t_sd = b3_mu0_t_sd,
                chains = 1)
```

---

### Results and Visualization

#### Pseudotime and Branch Inference

```{r pseudotime-inference, message=FALSE, warning=FALSE}
bt <- infer_pseudotime(bt)
bt <- infer_mu0(bt)
bt <- infer_k(bt)
bt <- infer_t0(bt)
bt <- branch_probability(bt)
bt <- calculate_branches(bt, limit_prob = 0.9)

# Pseudotime results
unname(bt$t)
bt$para_mu0
bt$para_k
bt$para_t0
```

#### Branch-Specific Gene Analysis

```{r branch-analysis, message=FALSE,results='hide', warning=FALSE}
# Extract branch-specific genes
t1 <- bt$all_branch_data[[1]]$t
expr1 <- data_counts[, bt$all_branch_data[[1]]$index]
filtered_genes1 <- filter_de_genes(t1, as.matrix(expr1), limit_cor = 0.3, limit_bf = 1e5)

t2 <- bt$all_branch_data[[2]]$t
expr2 <- data_counts[, bt$all_branch_data[[2]]$index]
filtered_genes2 <- filter_de_genes(t2, as.matrix(expr2), limit_cor = 0.45, limit_bf = 1e17)

t3 <- bt$all_branch_data[[3]]$t
expr3 <- data_counts[, bt$all_branch_data[[3]]$index]
filtered_genes3 <- filter_de_genes(t3, as.matrix(expr3), limit_cor = 0.4, limit_bf = 1e14)
```

#### Gene Dynamics Visualization




### Conclusion

This analysis demonstrates the ability of **BayesTraj** to identify and visualize lineage-specific gene expression dynamics across three major hematopoietic lineages.

# Technical info

```{r sess-info}
sessionInfo()
```
